### ISSUE TITLE
Testing Promises

### ISSUE THREAD

**Issue by oravecz** [shouldjs/promised#11]
What is the strategy for testing Promises? I have tried this with both Bluebird and Node built-in Promises. This passes because the promise doesn't resolve by the time the assertions are executed. I am probably misunderstanding how eventually and finally are used?

```
var Promise = require('bluebird') // or comment this out
require( 'should' )
require( 'should-promised' )

describe( 'sample test', function () {

    it('should recognize promises', function() {
        var p = new Promise( (resolve, reject) => {
            setTimeout( () => {
                reject('some reason')
            }, 2000)
        } )

        p.should.be.a.Promise()
        p.should.be.rejected()
        p.should.not.be.rejected()
        p.should.eventually.be.rejected()
        p.should.eventually.not.be.rejected()
        p.should.finally.be.rejected()
        p.should.finally.not.be.rejected()
    })
} );

```

**Comment by btd** [comment]
Most of should-promised methods returns promise as a return value. You
wrote just plain statements and does not deal with their return values
and/or done callback. Can you please take a look at tests of you will see
difference . I could not right examples anytime soon. Sorry.
28 Ноя 2015 г. 17:41 пользователь "Jim Cook" notifications@github.com
написал:

**Comment by oravecz** [comment]
But none of the tests actually deal with an asynchronous promise. They just immediately resolve or reject a value. I can add all of the checks I want to my test case and the assertions never throw an exception. For example:

```
it('should recognize promises', function*() {
    var p = new Promise( (resolve, reject) => {
        setTimeout( () => {
            reject('some reason')
        }, 2000)
    } )

    p.should.eventually.be.rejected().then.not.equal(/some reason/)
```

This test will succeed.

**Comment by btd** [comment]
You do not wait for promise to finish. You need return it from your test or use `done` callback. I am not sure if mocha starts support generators.

**Comment by oravecz** [comment]
I'm using co-mocha for generator support in testing, but even attempts to yield the promise before assertions do not work with this library. 

If I involve `done()` in my tests, I can't use this library because the assertion errors would prevent `done()` from being called. When using done(), I need to handle the assertions myself. I don't see how this library helps for the majority of promises used that are asynchronous. 

Here is an example that uses `done()` as an asynchronous test case. Can you show me how my example will be re-written using `should-promised`? 

```
    it( 'should handle async', function ( done ) {

        new Promise( ( resolve, reject ) => {
            setTimeout( () => {
                reject( 'some reason' )
            }, 1000 )
        } )
            .then( () => {
                // error condition
                done( 'failed to reject' )
            } )
            .catch( error => {
                if ( /some reason/.test( error ) ) {
                    // no error, expected result
                    done()
                } else {
                    done( 'rejected incorrectly' )
                }
            } )
    } )
```

**Comment by btd** [comment]
I am sorry but I am not able to write you example for several days  (I am
in hospital right now and writing from phone). If you would wait I will
help you rewriting your example

вс, 29 Ноя 2015, 21:44, Jim Cook notifications@github.com:

**Comment by oravecz** [comment]
No problem, sorry to hear it. Get well.

**Comment by oravecz** [comment]
Here is a slightly more compact version using a generator

```
    it( 'generator tested with should (no promised)', function *() {
        try {
            yield new Promise( ( resolve, reject ) => {
                setTimeout( () => {
                    reject( 'some reason' )
                }, 1000 )
            } )
            should.fail('should not succeed')
        } catch ( error ) {
            error.should.match(/some reason/)
        }
    } )
```

But this approach destroys the intention of asserting promises, and I don't see how I can apply it to any of the assertions from `should-promises`.

**Comment by oravecz** [comment]
I switched to using [chai-as-promised](http://chaijs.com/plugins/chai-as-promised), and it worked right away. I've been using the `shouldjs` library with the promised plugin, but it should be a seamless transition to chai. In fact, I'm not even sure any of my tests have to be changed.

```
    it( should work with chai-as-promised', function ( ) {
        return new Promise( ( resolve, reject ) => {
            setTimeout( () => {
                reject( 'some reason' )
            }, 1000 )
        } )
        .should.be.rejectedWith( /some reason/ )
    } )
```

**Comment by btd** [comment]
So at the end you did like I said in first my message. You returned promise
from test. Not sure about chai, but should.js returns in most methods
promise. So if in your last example you will use should promised result
should be the same
30 Ноя 2015 г. 3:44 пользователь "Jim Cook" notifications@github.com
написал:

**Comment by oravecz** [comment]
I'm not sure you noticed, but I never got it to work using should-promised. I had to switch to use chai-as-promised, and it worked right away. I'd be happy to look at your rewrite of my test case if you get a chance to post it.

**Comment by btd** [comment]
I just published 8.0.0 where i merged this part with main repo. Especially for you i added one test to show async promise (fyi all .then calls happen in next tick).
http://github.com/shouldjs/should.js/blob/871fd61d0ec22073a2887be96dda94a05a0cf33a/test/ext/promise.test.js#L142-L144

Let me know if you still have questions.
